### PR TITLE
feat: add macro to generate ToEncryptable trait

### PR DESCRIPTION
**Pull request by dracarys18**
## Type of Change

- [x] Enhancement

## Description

This PR adds a macro to generate ToEncryptable trait on different types

## Motivation and Context

This PR removes the effort of repetitively implementing ToEncryptable trait which is used for batch encryption/decryption in the encryption service by adding a macro which generates code upon deriving it

## How did you test it?

**THIS CANNOT BE TESTED ON SANDBOX**
- Compiler guided. Since this does not change any core flow
- Basic test of MerchantCreate and Payments for sanity

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code

### CODE REVIEW

**Review by ArjunKarthik — APPROVED** [review]

File: crates/router_derive/src/macros/to_encryptable.rs [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+        let temp_fields = struct_type.generate_struct_fields(&inner_types);
+        quote! {
+            pub struct #name {
+                #(#temp_fields,)*
+            }
+        }
+    });
+
+    // These implementations shouldn't be implemented Decrypted and DecryptedUpdate temp structs
+    // So skip the first two entries in the list
```
Comment (SanchithHegde) [inline crates/router_derive/src/macros/to_encryptable.rs:227]: Can we add a comment above the array declaration that states that `Decrypted` and `DecryptedUpdate` must be the first 2 members of the array?

File: crates/router_derive/src/macros/to_encryptable.rs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+    let inner_type = inner_types.first().map(|(_, ty)| ty).ok_or_else(|| {
+        syn::Error::new(
+            proc_macro2::Span::call_site(),
+            "Please use the macro with attribute #[encrypt] on the fields you want to encrypt",
+        )
+    })?;
+
+    let structs = struct_types.iter().map(|(suffix, struct_type)| {
+        let name = format_ident!("{}{}", suffix, struct_name);
```
Comment (SanchithHegde) [inline crates/router_derive/src/macros/to_encryptable.rs:217]: Is the `suffix` supposed to be attached after the `struct_name` or before? If it's before, should we consider calling it `prefix`?

File: crates/hyperswitch_domain_models/src/customer.rs [hunk 3]
```diff
@@ -113,16 +120,23 @@ impl super::behaviour::Conversion for Customer {
...
         Ok(Self {
             customer_id: item.customer_id,
             merchant_id: item.merchant_id,
             name: encryptable_customer.name,
-            email: encryptable_customer.email,
+            email: encryptable_customer.email.map(|email| {
+                let encryptable: Encryptable<Secret<String, pii::EmailStrategy>> = Encryptable::new(
+                    email.clone().into_inner().switch_strategy(),
+                    email.into_encrypted(),
+                );
```
Comment (SanchithHegde) [inline crates/hyperswitch_domain_models/src/customer.rs:137]: Why are we calling `switch_strategy()` here, can't we work with `Email` directly?

File: crates/router_derive/src/macros/to_encryptable.rs [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
```
Comment (SanchithHegde) [inline crates/router_derive/src/macros/to_encryptable.rs:219]: Nit:

```suggestion
    let structs = struct_types.iter().map(|(prefix, struct_type)| {
        let name = format_ident!("{}{}", prefix, struct_name);
```

**Review by SanchithHegde — DISMISSED** [review]

File: crates/router_derive/src/lib.rs [hunk 5]
```diff
@@ -694,3 +695,18 @@ pub fn flat_struct_derive(input: proc_macro::TokenStream) -> proc_macro::TokenSt
 
     proc_macro::TokenStream::from(expanded)
 }
+
+/// Generates the ToEncryptable trait for a type
+///
+/// This macro generates the Temperory structs which has the fields that needs to be encrypted
```
Comment (SanchithHegde) [inline crates/router_derive/src/lib.rs:701]: Nit: Typo.

```suggestion
/// This macro generates the temporary structs which has the fields that needs to be encrypted
```

**Review by ArjunKarthik — DISMISSED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**